### PR TITLE
[1.21] Select Music Event

### PR DESCRIPTION
**Pull request by Zepalesque**
What this does is adds an event for when `Minecraft#getSituationalMusic()` gets fired, allowing for modders to change/set the music based on whatever conditions they want (that are available from the client of course)

hopefully I did everything right, this is my first ever pull reqeust for either forge *or* neoforge, but I did follow the contributing instructions so hopefully it should be good

### CODE REVIEW

**Review by embeddedt — CHANGES_REQUESTED** [review]
We strive to keep the size of the patches themselves as small as possible. Instead of patching `Minecraft#getSituationalMusic`, you could replace the single call to it in `MusicManager` with `ClientHooks.getSituationalMusic()` and have that method call the original vanilla method, fire the event, and return the appropriate value. Then we no longer need a bunch of `return` -> `result =` patches.

**Comment by Zepalesque** [comment]
ah yeah, good idea, hadn’t thought of that

**Comment by Zepalesque** [comment]
ok, changes made

File: patches/net/minecraft/client/sounds/MusicManager.java.patch [hunk 1]
```diff
@@ -0,0 +1,25 @@
+--- a/net/minecraft/client/sounds/MusicManager.java
++++ b/net/minecraft/client/sounds/MusicManager.java
+@@ -25,6 +_,13 @@
+ 
+     public void tick() {
+         Music music = this.minecraft.getSituationalMusic();
++        Music posted = net.neoforged.neoforge.client.ClientHooks.selectMusic(music);
```
Comment (embeddedt) [inline patches/net/minecraft/client/sounds/MusicManager.java.patch]: Just reassign `music` here instead of making a second local, it avoids needing to patch `this.startPlaying` further down.

**Review by embeddedt — CHANGES_REQUESTED** [review]
Much better, I just have one comment

**Comment by Zepalesque** [comment]
ah yeah good idea, I’ll do this when get the chance

**Comment by ChampionAsh5357** [comment]
Honestly, I have no idea how this will work in correlation to multiple mods. For example, if a mod played situational music based upon being in a given structure while a different mod played situational music based on an item the player was wearing on their head, which one would take precedence?

I'm not against the PR, but I think there should be some ordering that should be either defined or explained with some custom system or using the priority of the event listener (e.g., HIGHEST for helmet, LOW for in a structure, LOWEST for a biome/dimension. Other ideas are welcome.

Also, I think the music that should be present in the event is the current music, not the one that Mojang is going to take over after yours.

**Comment by Zepalesque** [comment]
hmm yeah I can probably add some documentation stuff for the event priority stuff, good idea

for the other part, I’m not 100% sure what you mean, but if you mean that you should be able to change the `currentMusic` field, then there are a few reasons I didn’t do that, mainly because letting you set that could break stuff since that’s just used to track the current music, so changing it would likely lead to things like overlapping music tracks, plus you can access it if necessary for some reason or another anyway through `Minecraft.getInstance().musicManager.currentMusic`

You can also still have the music track instantly start playing as well and stop existing music, if that’s what you meant, you just have to set the Music so that `replaceCurrentMusic()` returns true (can’t remember the exact field name atm)

The specific use case this was initially meant for was for the Aether mod, which overrides creative music in its biomes so that it plays the biome music instead, plus one of its addons that additionally adds nighttime music tracks for the dimension
These are supposed to not instantly replace the existing music, but instead they are supposed to play once the previous music is done and the delay timer ends (for the nighttime music tracks at least, the others are just made the default for the dimension of that makes sense), similar to vanilla music, just allowing a more dynamic system, while also allowing for there to still be other uses

Previously the Aether had used it’s own entirely new music manager, but I figured out that it wouldn’t really be necessary to have a separate one if there was a way to modify what music gets chosen based on the situation

heck this got long, whoops lol, hopefully it’s not too convoluted :joy:

**Comment by ChampionAsh5357** [comment]
I don't mean you should change the `currentMusic` field, rather that the information makes more sense in terms of the event context. The situational music from Mojang just indicates what the default music would be playing. If we know the event logic will play its own sound, it might not make sense to execute `Minecraft#getSituationalMusic`. The current music is more useful as it indicates what the custom music will be overriding if it chooses to do so.

**Comment by Zepalesque** [comment]
The issue with that is that if you wanted to somehow utilize the original default music track for the situation when choosing the new potential music track, you wouldn’t be able to if it instead gave you the current playing music (for instance, using a map to switch to a different music track if a certain one would’ve been played plus a certain condition, such as it being nighttime)
Not only that, but it could also cause discrepancy between the situation and the music, as if you used the current music for the same purpose instead, that might not be the default one based on the current situation, and it would give the incorrect music
it’s also meant to be able to be chained together, so that if you have the event fire after another mod’s event you are given both the original music and the one that the other mod set it to

pretty much, the reason it calls `Minecraft#getSituationalMusic` is so that it can be used as a variable for the modder when they are choosing the music they want to have chosen instead
hopefully that makes sense and isn’t just a big jumble of word soup

**Comment by TelepathicGrunt** [comment]
Note for my Bumblezone mod, I have a structure that should stop all other music when player is in it and plays the structure’s own music. Then there is an arena event in the structure that can be triggered that will stop the structure music and play the arena event music. 

Which means my use case is one where I’m going to want to override everyone else’s music. At least to me, I feel like I want high priority. Is there a reason why a helmet music should override my event music?

**Comment by Zepalesque** [comment]
most likely what you’d want to do is have the structure music be high priority, and then the event music *higher* priority, and have both `replaceCurrentMusic`
not sure what you mean by helmet music tho

**Comment by ChampionAsh5357** [comment]
> pretty much, the reason it calls Minecraft#getSituationalMusic is so that it can be used as a variable for the modder when they are choosing the music they want to have chosen instead

I understand, though I feel as though `Minecraft#getSituationalMusic` also falls into that category as whatever it outputs would be the currently playing music, but eh. If you believe that it doesn't make sense, I'll leave it at that since I don't have a solid enough argument for or against it.

> Which means my use case is one where I’m going to want to override everyone else’s music. At least to me, I feel like I want high priority. Is there a reason why a helmet music should override my event music?

My general thought process was that things closer to the player in context would take higher priority, since that would be what the player is focusing on. However, it's up to the modders to determine which priority makes sense for their mod. I was just using the helmet as an example.

**Comment by Zepalesque** [comment]
ah ok I see, I didn’t see the part about having an item on the players head before lol so I was confused when telepathicgrunt mentioned helmets lol

but yeah music keeps playing even if it’s not the one that would’ve been picked, so if you are in one biome with one music, which is playing, and you go to another with different music then it won’t just instantly stop the music so it won’t necessarily be the same

hmm, semi-unrelated to this *specific* conversation, but I just realized an issue that changing it so it reassigns the music variable rather than having a separate posted variable would resolve, so that’s a good thing

**Comment by Zepalesque** [comment]
ok, changed it so it reassigns the `music` variable along with having more in-depth documentation

**Review by Shadows-of-Fire — DISMISSED** [review]

File: patches/net/minecraft/client/sounds/MusicManager.java.patch [hunk 2]
```diff
@@ -0,0 +1,14 @@
+--- a/net/minecraft/client/sounds/MusicManager.java
++++ b/net/minecraft/client/sounds/MusicManager.java
+@@ -24,7 +_,10 @@
+     }
+ 
+     public void tick() {
+-        Music music = this.minecraft.getSituationalMusic();
++        Music music = net.neoforged.neoforge.client.ClientHooks.selectMusic(this.minecraft.getSituationalMusic());
++        if (music == null) {
```
Comment (ChampionAsh5357) [inline patches/net/minecraft/client/sounds/MusicManager.java.patch:9]: What is null supposed to do here? Is this stopping the music or keeping the current music playing?

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]

File: patches/net/minecraft/client/sounds/MusicManager.java.patch (hunk 2 above)
Comment (Zepalesque) [reply patches/net/minecraft/client/sounds/MusicManager.java.patch:9]: yeah, if the music is set to null by the event, it’s meant to cancel the music from potentially playing

Comment (Zepalesque) [reply patches/net/minecraft/client/sounds/MusicManager.java.patch:9]: the music only should be *stopped and replaced* if the Music object specifies it, as in if `Music#replaceCurrentMusic` is true
If the music is set to null, it will not try to play the next music track
Otherwise, and if the music is *not* meant to replace the current music, it will use that music track for the cooldown logic (such as choosing how long until the next music track should play), and if the already-playing music has stopped and the cooldown has finished, the music set by the event is the one that plays.

Comment (ChampionAsh5357) [reply patches/net/minecraft/client/sounds/MusicManager.java.patch:9]: Ok, what does 'cancel the music from potentially playing mean'? Does that mean stop music altogether or stop any other modifications to the song from being played, because both have their issues.

Returning `null` just keeps the current song playing until it eventually stops. Then, no song will be played **at all** until some song is returned and the `nextSongDelay` finishes its countdown, which could be a long realm of silence. The quick end of the logic seems to have weird behavior.

I think there should be a way to stop all music, and that the `nextSongDelay` isn't blocked from the music being stopped.

**Comment by Zepalesque** [comment]
Ok, made the changes requested

File: patches/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java.patch [hunk 3]
```diff
@@ -59,12 +59,10 @@
      }
  
      @Override
-@@ -125,7 +_,13 @@
-     public void handleConfigurationFinished(ServerboundFinishConfigurationPacket p_294283_) {
+@@ -126,6 +_,12 @@
          PacketUtils.ensureRunningOnSameThread(p_294283_, this, this.server);
          this.finishCurrentTask(JoinWorldTask.TYPE);
--        this.connection.setupOutboundProtocol(GameProtocols.CLIENTBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess())));
-+        this.connection.setupOutboundProtocol(GameProtocols.CLIENTBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess(), this.connectionType)));
```
Comment (ChampionAsh5357) [inline patches/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java.patch]: Mistake during merge?

File: patches/net/minecraft/world/item/ItemStack.java.patch
Comment (ChampionAsh5357) [inline patches/net/minecraft/world/item/ItemStack.java.patch:1]: Mistake during merging?

File: src/main/java/net/neoforged/neoforge/client/event/SelectMusicEvent.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+     * Changes the situational music. If this is set to {@code null}, any currently playing music will be cancelled.<br>
+     * If this <i>was</i> {@code null} but on the next tick isn't, the muisc given will be immediately played.<br>
+     * <br>
+     * Note that if {@link #isCanceled()} is {@code true}, this method will do nothing, as it is meant to override other musics.
+     */
+    public void setMusic(@Nullable Music newMusic) {
+        if (!this.isCanceled()) {
+            this.music = newMusic;
+        }
+    }
```
Comment (ChampionAsh5357) [inline src/main/java/net/neoforged/neoforge/client/event/SelectMusicEvent.java]: So this is not what I meant. When I said cancelled, I meant that the modder chooses whether to cancel the event. If the modder cancels the event, then no other listeners will be invoked, meaning that whatever this listener selects will be the music played. This wrapping check is unnecessary. The only thing that may be useful is two versions of the set method where one calls `setCanceled` for you.

File: patches/net/minecraft/client/sounds/MusicManager.java.patch [hunk 5]
```diff
@@ -0,0 +1,43 @@
+--- a/net/minecraft/client/sounds/MusicManager.java
++++ b/net/minecraft/client/sounds/MusicManager.java
+@@ -24,22 +_,28 @@
+     }
+ 
+     public void tick() {
+-        Music music = this.minecraft.getSituationalMusic();
++        Music music = net.neoforged.neoforge.client.ClientHooks.selectMusic(this.minecraft.getSituationalMusic(), this.currentMusic);
```
Comment (ChampionAsh5357) [inline patches/net/minecraft/client/sounds/MusicManager.java.patch:8]: This patch is generally too large. The only thing you really need to do is provide a separate if check that if `music` is null and `currentMusic` is not null to stop the current music and set the next song delay to 0. You can return after this.

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]
Looking good. A few changes because you may have misunderstood what I meant in discord.

File: patches/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java.patch (hunk 3 above)
Comment (Zepalesque) [reply patches/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java.patch]: oh heck yeah oops lol
I

File: src/main/java/net/neoforged/neoforge/client/event/SelectMusicEvent.java (hunk 4 above)
Comment (Zepalesque) [reply src/main/java/net/neoforged/neoforge/client/event/SelectMusicEvent.java]: hmm, don't the other listeners still get run even if it was cancelled?

File: patches/net/minecraft/client/sounds/MusicManager.java.patch (hunk 5 above)
Comment (Zepalesque) [reply patches/net/minecraft/client/sounds/MusicManager.java.patch:8]: ok, fixed this one

File: src/main/java/net/neoforged/neoforge/client/event/SelectMusicEvent.java (hunk 4 above)
Comment (Zepalesque) [reply src/main/java/net/neoforged/neoforge/client/event/SelectMusicEvent.java]: nevermind, apparently I was wrong about this

Comment (Zepalesque) [reply src/main/java/net/neoforged/neoforge/client/event/SelectMusicEvent.java]: ok, fixed it

File: patches/net/minecraft/client/sounds/MusicManager.java.patch [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+             if (!music.getEvent().value().getLocation().equals(this.currentMusic.getLocation()) && music.replaceCurrentMusic()) {
+                 this.minecraft.getSoundManager().stop(this.currentMusic);
+@@ -37,10 +_,15 @@
+             }
+         }
+ 
++        if (music == null) {
++            return;
++        }
++
```
Comment (ChampionAsh5357) [inline patches/net/minecraft/client/sounds/MusicManager.java.patch]: I think instead of this, it would make sense to just layer the if statement of current music inside of the is null check

File: patches/net/minecraft/client/sounds/MusicManager.java.patch [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+         this.nextSongDelay = Math.min(this.nextSongDelay, music.getMaxDelay());
+         if (this.currentMusic == null && this.nextSongDelay-- <= 0) {
+             this.startPlaying(music);
+         }
++
```
Comment (ChampionAsh5357) [inline patches/net/minecraft/client/sounds/MusicManager.java.patch]: Unnecessary change.

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]
Just some patch issues, but other than that, looks good.

File: patches/net/minecraft/client/sounds/MusicManager.java.patch (hunk 6 above)
Comment (Zepalesque) [reply patches/net/minecraft/client/sounds/MusicManager.java.patch]: true, good point

File: patches/net/minecraft/client/sounds/MusicManager.java.patch (hunk 7 above)
Comment (Zepalesque) [reply patches/net/minecraft/client/sounds/MusicManager.java.patch]: whoops lol

**Comment by Zepalesque** [comment]
okay, fixed those as well

**Review by ChampionAsh5357 — DISMISSED** [review]
Looks fine now.

**Review by TelepathicGrunt — APPROVED** [review]

**Review by embeddedt — APPROVED** [review]
